Upgrade to TypeScript 2.5

alexeagle opened the pull request:
- update to TypeScript 2.5
- point the 2.4 typings test at the previous typescript version, so we
  don't break it accidentally
- update to latest TypeScript 2.5 compatible Bazel rules
- move .bazelrc to tools/bazel.rc per https://docs.bazel.build/versions/master/best-practices.html#bazelrc

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?
```
[ ] Yes
[ ] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

alexeagle committed:
build: Upgrade to TypeScript 2.5

- update to TypeScript 2.5
- point the 2.4 typings test at the previous typescript version, so we
  don't break it accidentally
- widen the peerDeps from Angular packages that depend on TypeScript
- update to latest TypeScript 2.5 compatible Bazel rules
- move .bazelrc to tools/bazel.rc per https://docs.bazel.build/versions/master/best-practices.html#bazelrc

mary-poppins commented:
You can preview a1b7ab4 at https://pr20175-a1b7ab4.ngbuilds.io/.

mary-poppins commented:
You can preview 1a9242e at https://pr20175-1a9242e.ngbuilds.io/.

mary-poppins commented:
You can preview a803062 at https://pr20175-a803062.ngbuilds.io/.

IgorMinar reviewed (approved):
Please fix the commit message format. otherwise lgtm supposing that ci passes.

alexeagle commented:
@chuckjaz I hear that we need a single tsickle version that works with all supported typescript versions. Tobias says you have an idea for that.

mary-poppins commented:
You can preview 5cc5ef3 at https://pr20175-5cc5ef3.ngbuilds.io/.

mary-poppins commented:
You can preview 238307a at https://pr20175-238307a.ngbuilds.io/.

mary-poppins commented:
You can preview 3f47daa at https://pr20175-3f47daa.ngbuilds.io/.

chuckjaz commented on packages/compiler-cli/src/transformers/compiler_host.ts:
```diff
@@ -113,7 +113,7 @@ export class TsCompilerAotCompilerTypeCheckHostAdapter implements ts.CompilerHos
       },
       fileExists: (filePath) => this.originalFileExists(filePath),
-      readFile: (filePath) => this.context.readFile(filePath),
+      readFile: (filePath) => this.context.readFile(filePath) !,
```
Can you add a clean-up TODO comment here and other places that are similar below?

alexeagle replied:
Sure, added an `assert` helper at the top of the file with a TODO for you. I don't want to change the behavior to throw in this PR.

chuckjaz reviewed (approved):
One nit and a request to add a typing_test_ts25.

mary-poppins commented:
You can preview 4e71a80 at https://pr20175-4e71a80.ngbuilds.io/.

mary-poppins commented:
You can preview a79abe0 at https://pr20175-a79abe0.ngbuilds.io/.

mary-poppins commented:
You can preview 82d87a2 at https://pr20175-82d87a2.ngbuilds.io/.

mary-poppins commented:
You can preview b8e4d39 at https://pr20175-b8e4d39.ngbuilds.io/.

mary-poppins commented:
You can preview 100bebe at https://pr20175-100bebe.ngbuilds.io/.

mary-poppins commented:
You can preview c66e417 at https://pr20175-c66e417.ngbuilds.io/.

alexeagle committed:
fix(bazel): adjust mock of tsconfig for ng_module rule unit test

Due to https://github.com/bazelbuild/rules_typescript/commit/7f33f01b72e61f8629d129c46998d96f5bd1977e

mary-poppins commented:
You can preview 480014d at https://pr20175-480014d.ngbuilds.io/.

alexeagle committed:
fix: Update test code to type-check under TS 2.5

mary-poppins commented:
You can preview d729cc0 at https://pr20175-d729cc0.ngbuilds.io/.

mary-poppins commented:
You can preview 2a3ccc6 at https://pr20175-2a3ccc6.ngbuilds.io/.

mary-poppins commented:
You can preview ef3c872 at https://pr20175-ef3c872.ngbuilds.io/.

mary-poppins commented:
You can preview a8ba3ce at https://pr20175-a8ba3ce.ngbuilds.io/.

mary-poppins commented:
You can preview 463872a at https://pr20175-463872a.ngbuilds.io/.

alexeagle commented:
Note, if you call `Injector.get(Token)` with a Token with static members, you'll get the wrong type, see https://github.com/Microsoft/TypeScript/issues/20102
